File: recursion_and_dynamic_programming/intmul/intmul.go
package intmul

// Mul multiples a and b.
func Mul(a, b int) int {
	ans := 0
	for i := 0; i < b; i++ {
		ans += a
	}
	return ans
}

// Mul2 multiples a and b.
func Mul2(a, b int) int {
	multiplicand, multiplier := a, b
	if multiplicand < multiplier {
		multiplicand, multiplier = multiplier, multiplicand
	}
	ans := 0
	for i := 0; i < multiplier; i++ {
		ans += multiplicand
	}
	return ans
}

// Mul3 multiplies a and b.
func Mul3(a, b int) int {
	smaller, bigger := a, b
	if smaller > bigger {
		smaller, bigger = bigger, smaller
	}
	return mul3(smaller, bigger)
}

func mul3(smaller, bigger int) int {
	if smaller == 0 {
		return 0
	} else if smaller == 1 {
		return bigger
	}
	next := smaller >> 1
	side := mul3(next, bigger)
	if smaller%2 == 1 {
		return side + side + bigger
	}
	return side + side
}

File: recursion_and_dynamic_programming/intmul/intmul.py
def intmul(a: int, b: int) -> int:
    def intmul_innter(a: int, b: int, s: int) -> int:
        if b == 0:
            return 0
        s += intmul_innter(a, b-1, s)
        return s+a
    return intmul_innter(a, b, 0)


def intmul2(a: int, b: int) -> int:
    s = 0
    for _ in range(b, 0, -1):
        s += a
    return s


def intmul3(a: int, b: int) -> int:
    def intmul_inner(a: int, b: int) -> int:
        for _ in range(b, 0, -1):
            yield a
    return sum(intmul_inner(a, b))


def intmul4(a: int, b: int) -> int:
    def intmul_inner(smaller: int, bigger: int) -> int:
        if smaller == 0:
            return 0
        if smaller == 1:
            return bigger
        q = smaller >> 1
        side = intmul_inner(q, bigger)
        if smaller % 2 == 1:
            return side + side + bigger
        else:
            return side + side
    smaller, bigger = (a, b) if a < b else (b, a)
    return intmul_inner(smaller, bigger)


for i in range(10):
    print("intmul: ", intmul(2, i))
    print("intmul2:", intmul2(20, i))
    print("intmul3:", intmul3(200, i))
    print("intmul4:", intmul4(2000, i))
